src/cpufreq.c

```diff
@@ -82,9 +132,67 @@ static int cpufreq_read(void) {
     /* convert kHz to Hz */
     v.gauge *= 1000.0;
 
-    cpufreq_submit(i, v);
+    cpufreq_submit(i, "cpufreq", NULL, &v);
+
+    if (report_p_stats) {
+      /* Read total transitions for cpu frequency */
+      snprintf(filename, sizeof(filename),
+               "/sys/devices/system/cpu/cpu%d/cpufreq/stats/total_trans", i);
+      if (parse_value_file(filename, &v, DS_TYPE_COUNTER) != 0) {
+        ERROR("cpufreq plugin: Reading \"%s\" failed.", filename);
+        continue;
+      }
+      cpufreq_submit(i, "counter", "transitions", &v);
+
+      /* Determine percentage time in each state for cpu during previous
+       * interval. */
+      snprintf(filename, sizeof(filename),
+               "/sys/devices/system/cpu/cpu%d/cpufreq/stats/time_in_state", i);
+
+      FILE *fh = fopen(filename, "r");
+      if (fh == NULL) {
+        ERROR("cpufreq plugin: Reading \"%s\" failed.", filename);
+        continue;
+      }
+
+      int j = 0;
+      cdtime_t now = cdtime();
+      char buffer[DATA_MAX_NAME_LEN];
+
+      while (fgets(buffer, sizeof(buffer), fh) != NULL) {
+        unsigned long long time;
+        char state[DATA_MAX_NAME_LEN];
+
+        /*
+         * State time units is 10ms. To get rate of seconds per second
+         * we have to divide by 100. To get percents we have to multiply it
+         * by 100 back. So, just use parsed value directly.
+         */
+        if (!sscanf(buffer, "%s%llu", state, &time)) {
```

Comment:
*sscanf()* doesn't know the size of `state` and is prone to buffer overflows. Please don't use it.

Reply:
Replaced %s with %u.